docs/error-handling.md

```diff
@@ -0,0 +1,207 @@
+Error handling
+==============
+
+Error Kinds
+-----------
+The following kinds of error would appears in the codebase:
+
+### Panic
+Panics are produced with Go's built-in `panic` function. It is used for:
+- impossible conditions (e.g. exhaustive enum matching)
+- unrecoverable server error (e.g. invalid server/teanant configuration)
+
+`panic` should be called with a `string` or `error` argument. If the argument
+is a string, it should be in format `<module>: <message>`. For example:
+`sso: unknown provider type: example`.
+
+Panics are recovered by a middleware. Recovered panics would be logged and
+produce a 500 response if needed.
+
+### Plain error
+Plain errors are produced by error constructing functions (e.g. `errors.New`)
+or returned from dependency library. Unless otherwise required, this kind of
```

Comment:
s/dependency/dependent

Reply:
I think dependent has incorrect meaning: 'dependent library' would mean the library that depends on it.

Reply:
Both of them have the same meaning.

Both mean library which the application depends on.

Reply:
I don't think so? For example, npm use 'dependency' to indicate the package it depends on, and use 'dependent' to indicate the package that depends on it.